File: Tutorials/Codes/Modern_Control_Course/Chapter30/Lesson4/Chapter30_Lesson4.py
# Chapter30_Lesson4.py
# Integrated case study: mass-spring-damper model to state-feedback controller
# Libraries: NumPy for linear algebra. SciPy is optional for comparison with place_poles.

import math
import numpy as np


def controllability_matrix(A: np.ndarray, B: np.ndarray) -> np.ndarray:
    """Return [B, AB] for a two-state SISO model."""
    return np.column_stack((B, A @ B))


def state_feedback_from_specs(m: float, b: float, k: float, zeta: float, settling_time: float):
    """Compute K = [k1, k2] by matching the desired second-order polynomial."""
    omega_n = 4.0 / (zeta * settling_time)
    k1 = m * omega_n**2 - k
    k2 = 2.0 * zeta * omega_n * m - b
    return np.array([[k1, k2]], dtype=float), omega_n


def feedforward_gain(A: np.ndarray, B: np.ndarray, C: np.ndarray, K: np.ndarray) -> float:
    """Nbar = -1/(C (A-BK)^(-1) B), valid when the scalar denominator is nonzero."""
    Acl = A - B @ K
    denom = C @ np.linalg.solve(Acl, B)
    return float(-1.0 / denom[0, 0])


def rk4_step(f, t, x, h):
    k1 = f(t, x)
    k2 = f(t + 0.5 * h, x + 0.5 * h * k1)
    k3 = f(t + 0.5 * h, x + 0.5 * h * k2)
    k4 = f(t + h, x + h * k3)
    return x + (h / 6.0) * (k1 + 2.0 * k2 + 2.0 * k3 + k4)


def simulate_closed_loop(A, B, C, K, nbar, r=1.0, t_final=6.0, dt=0.002):
    Acl = A - B @ K
    Bcl = B * nbar

    def f(t, x):
        return Acl @ x + Bcl.flatten() * r

    steps = int(t_final / dt)
    t = np.zeros(steps + 1)
    x = np.zeros((steps + 1, 2))
    y = np.zeros(steps + 1)
    u = np.zeros(steps + 1)
    for i in range(steps):
        t[i + 1] = t[i] + dt
        y[i] = float(C @ x[i])
        u[i] = float(-K @ x[i].reshape(-1, 1) + nbar * r)
        x[i + 1] = rk4_step(f, t[i], x[i], dt)
    y[-1] = float(C @ x[-1])
    u[-1] = float(-K @ x[-1].reshape(-1, 1) + nbar * r)
    return t, x, y, u


def main():
    m = 1.2       # kg
    b = 0.8       # N*s/m
    k = 3.0       # N/m
    zeta = 0.70
    settling_time = 2.0

    A = np.array([[0.0, 1.0], [-k / m, -b / m]])
    B = np.array([[0.0], [1.0 / m]])
    C = np.array([[1.0, 0.0]])
    D = np.array([[0.0]])

    Wc = controllability_matrix(A, B)
    print("A =\n", A)
    print("B =\n", B)
    print("Controllability matrix =\n", Wc)
    print("rank(Wc) =", np.linalg.matrix_rank(Wc))

    K, omega_n = state_feedback_from_specs(m, b, k, zeta, settling_time)
    nbar = feedforward_gain(A, B, C, K)
    Acl = A - B @ K

    print("omega_n =", omega_n)
    print("K =", K)
    print("Nbar =", nbar)
    print("closed-loop poles =", np.linalg.eigvals(Acl))

    t, x, y, u = simulate_closed_loop(A, B, C, K, nbar, r=1.0)
    print("final output y(t_final) =", y[-1])
    print("maximum absolute control input =", np.max(np.abs(u)))

    # Save a small CSV for plotting in any external tool.
    data = np.column_stack((t, x[:, 0], x[:, 1], y, u))
    np.savetxt(
        "Chapter30_Lesson4_response.csv",
        data,
        delimiter=",",
        header="t,position,velocity,output,control",
        comments="",
    )


if __name__ == "__main__":
    main()

File: Tutorials/Codes/Modern_Control_Course/Chapter30/Lesson4/Chapter30_Lesson4.m
% Chapter30_Lesson4.m
% Integrated case study: mass-spring-damper model to state-feedback controller
% Libraries: base MATLAB for matrix work; Control System Toolbox for ss/place/lsim;
% Simulink section is optional and creates a closed-loop State-Space block.

clear; clc; close all;

m = 1.2;      % kg
b = 0.8;      % N*s/m
k = 3.0;      % N/m
zeta = 0.70;
Ts = 2.0;
omega_n = 4/(zeta*Ts);

A = [0 1; -k/m -b/m];
B = [0; 1/m];
C = [1 0];
D = 0;

Wc = [B A*B];
fprintf('rank(Wc) = %d\n', rank(Wc));

% Manual pole-matching formula for this second-order plant.
k1 = m*omega_n^2 - k;
k2 = 2*zeta*omega_n*m - b;
K_manual = [k1 k2];

% Optional toolbox verification using place.
desiredPoles = roots([1 2*zeta*omega_n omega_n^2]);
if exist('place', 'file') == 2
    K_place = place(A, B, desiredPoles);
else
    K_place = K_manual;
end

K = K_manual;
Acl = A - B*K;
Nbar = -1/(C*(Acl\B));
Bcl = B*Nbar;

fprintf('K_manual = [%g %g]\n', K_manual(1), K_manual(2));
fprintf('K_place  = [%g %g]\n', K_place(1), K_place(2));
fprintf('Nbar = %g\n', Nbar);
disp('closed-loop poles ='); disp(eig(Acl));

% Simulate the closed-loop reference response.
t = 0:0.002:6;
r = ones(size(t));
if exist('ss', 'file') == 2 && exist('lsim', 'file') == 2
    sys_cl = ss(Acl, Bcl, C, D);
    y = lsim(sys_cl, r, t);
else
    f = @(tt, x) Acl*x + Bcl;
    [t_ode, x_ode] = ode45(f, t, [0; 0]);
    y = x_ode*C.';
    t = t_ode;
end

figure;
plot(t, y, 'LineWidth', 1.5); grid on;
xlabel('time (s)'); ylabel('position output');
title('Chapter 30 Lesson 4: closed-loop step response');

% Optional Simulink model for closed-loop state-space response.
% It uses the already-designed closed-loop matrices Acl and Bcl.
try
    if exist('new_system', 'file') == 4
        model = 'Chapter30_Lesson4_Simulink';
        if bdIsLoaded(model)
            close_system(model, 0);
        end
        new_system(model);
        add_block('simulink/Sources/Step', [model '/reference_step'], 'Position', [50 80 100 110]);
        add_block('simulink/Continuous/State-Space', [model '/closed_loop_state_space'], ...
            'A', mat2str(Acl), 'B', mat2str(Bcl), 'C', mat2str(C), 'D', mat2str(D), ...
            'Position', [170 68 310 122]);
        add_block('simulink/Sinks/Scope', [model '/scope'], 'Position', [380 75 430 115]);
        add_line(model, 'reference_step/1', 'closed_loop_state_space/1');
        add_line(model, 'closed_loop_state_space/1', 'scope/1');
        set_param(model, 'StopTime', '6');
        save_system(model);
        fprintf('Simulink model saved: %s.slx\n', model);
    end
catch ME
    fprintf('Simulink model creation skipped: %s\n', ME.message);
end
